src/crab/cfg.hpp

```diff
@@ -425,6 +454,18 @@ class cfg_t final {
         return labels;
     }
 
+    [[nodiscard]]
+    std::vector<label_t> potential_loop_headers() const {
+        std::vector<label_t> res;
+        res.reserve(m_blocks.size());
+        for (const auto& label : labels()) {
+            if (get_node(label).has_backwards_branch()) {
+                res.emplace_back(label);
+            }
+        }
+        return res;
+    }
```

Comment:
@caballa what do you think?

Reply:
Why not wto of the CFG and get the head of each component?

Reply:
That would be a fairly major rewrite. The sequence is currently:
1) [Prepare CFG](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/asm_cfg.cpp#L361C1-L383C2)
1.1) [Convert instructions to CFG](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/asm_cfg.cpp#L364C5-L364C66)
1.2) [Add assertions to CFG](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/asm_cfg.cpp#L367)
1.3) [Simplify the CFG](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/asm_cfg.cpp#L379)
2) [Gather report](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab_verifier.cpp#L164C1-L182C2)
2.1) [Setup entry conditions](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab_verifier.cpp#L173)
2.2) [Forward analysis](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab_verifier.cpp#L173)
2.2.1) [Build WTO](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab/fwd_analyzer.cpp#L104)
2.2.2) [Add increment loop counter to loop heads](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab/fwd_analyzer.cpp#L125C4-L136C6)
2.2.3) [Transform pre to post](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab/fwd_analyzer.cpp#L137C1-L141C58)
3) [Generate report](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab_verifier.cpp#L175)
3.1) [Check the assertions against the invariants at each label](https://github.com/vbpf/ebpf-verifier/blob/076ae86cfa17c55d06b93dcfc2067405469cca44/src/crab_verifier.cpp#L65C1-L85C18)

This change injects the IncrementLoopCounter into the deterministic CFG so that the assertions can be built from them in step 1.2 and at that point we haven't built the WTO, hence using the heuristic of backwards loops.

I thought the WTO was built from the simplified non-deterministic CFG.

So this would require either:
1) Using loop heuristics instead of WTO.
2) Building a temporary simplified non-deterministic CFG then building the WTO, just to add the assertions.
3) Adding the loop counter assertions later on during the analysis phases.
4) Walk the invariants after analysis and find the highest loop counter.

I initially went with option 4 (but folks pointed out that this was kind of a hack), then 3, but folks here suggested I should add the assertions during the normal explicate_assertions pass, but that is prior to building the WTO, hence using the heuristics instead of the WTO, hence option 1.

I am not confident enough in my understanding to take on the major overhaul that would be required to have the WTO at explicate assertions time without breaking something in a subtle manner.

Reply:
Thanks for your detailed explanation.
But we can compute WTO twice. You can build it after 1.1 and then again 2.2.1.
I don't see why it wouldn't work and performance should be fine since WTO is very cheap to compute.

Reply:
Switched over to using WTO to visit each loop head to either insert the increment loop counter instruction or initialize the loop counter variable.